p2p/peer_behaviour.go

```diff
@@ -0,0 +1,140 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrorPeerBehaviour are types of observable erroneous peer behaviours.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
```

Comment:
How do reactors specify reactor specific values for this, and should they need to? We can try to capture the common things, like timeouts and invalid messages, but presumably we may want to distinguish between those happening in the SecretConnection/MConn from those happening in specific reactors.

Reply:
Initially the reason for the introduction of the behaviour is to avoid disparate places in the codebase stopping a peer for a non-formalised error. We need to enumerate all possible errors in one place, ideally the place which decides if the error is severe enough to punish a peer. If we strive for a reactor centric definition of those errors we end up with the same kid of dynamic glue we currently try to reduce.

Do you have ideas how it could be expressed?

Reply:
I don't think we can avoid having some reactor-centric feature here without losing significant functionality or over-coupling our domains. 

At the least, we could standardize where in the respective reactor packages such errors are defined. For instance it could be part of channel registration: ie https://github.com/tendermint/tendermint/blob/develop/consensus/reactor.go#L125

Reply:
What if we move this to a higher level package as mentioned in
[#3552 comment](https://github.com/tendermint/tendermint/pull/3552#discussion_r280941065)? We could then follow a similar pattern as specified in [adr-30](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-030-consensus-refactor.md) to provide reactor-centric features:

Implementation: `behaviour/peer_behaviour.go:`
```go
struct PeerBehaviour interface {}

type PeerBehaviourVote struct {...}
...

type PeerBehaviourReporter interface {
	Report(peerID p2p.ID, behaviour PeerBehaviour) error
}
...
func (spbr *PeerBehaviourReporterXXX)Report(peerID ID, behaviour PeerBehaviour) error {
	...
	switch behaviour.(type) {
	case PeerBehaviourVote
        ..
		break
	...
	}
}
```

Example `consensus/reactor.go`:
```go
...
peerReporter.Report(peerID, PeerBehaviourVote{...})
```

Thoughts?